bindinfo/handle.go

```diff
@@ -838,57 +837,111 @@ func GenerateBindSQL(ctx context.Context, stmtNode ast.StmtNode, planHint string
-			restoreCtx.DefaultDB = defaultDB
-			err := n.With.Restore(restoreCtx)
-			if err != nil {
-				logutil.BgLogger().Debug("[sql-bind] restore SQL failed", zap.Error(err))
-				return ""
+		blockStartOffset = 1
+		afterProcessFunc = func(bindSQL string) string {
+			var selectIdx int
+			if n.With != nil {
+				selectIdx = strings.Index(bindSQL, "WITH")
+			} else {
+				selectIdx = strings.Index(bindSQL, "SELECT")
+			}
+			// Remove possible `explain` prefix.
+			return bindSQL[selectIdx:]
+		}
+	case *ast.SetOprStmt:
+		blockStartOffset = 1
+		afterProcessFunc = func(bindSQL string) string {
+			var selectIdx int
+			if n.With != nil {
+				selectIdx = strings.Index(bindSQL, "WITH")
+			} else {
+				selectIdx = strings.Index(bindSQL, "SELECT")
 			}
-			withEnd := withIdx + len(withSb.String())
-			tmp := strings.Replace(bindSQL[withEnd:], "SELECT", fmt.Sprintf("SELECT /*+ %s*/", planHint), 1)
-			return strings.Join([]string{bindSQL[withIdx:withEnd], tmp}, "")
+			// Remove possible `explain` prefix.
+			return bindSQL[selectIdx:]
 		}
-		selectIdx = strings.Index(bindSQL, "SELECT")
-		// Remove possible `explain` prefix.
-		bindSQL = bindSQL[selectIdx:]
-		return strings.Replace(bindSQL, "SELECT", fmt.Sprintf("SELECT /*+ %s*/", planHint), 1)
 	case *ast.InsertStmt:
-		insertIdx := int(0)
-		if n.IsReplace {
-			insertIdx = strings.Index(bindSQL, "REPLACE")
-		} else {
-			insertIdx = strings.Index(bindSQL, "INSERT")
+		blockStartOffset = 1
+		afterProcessFunc = func(bindSQL string) string {
+			var insertIdx int
+			if n.IsReplace {
+				insertIdx = strings.Index(bindSQL, "REPLACE")
+			} else {
+				insertIdx = strings.Index(bindSQL, "INSERT")
+			}
+			// Remove possible `explain` prefix.
+			return bindSQL[insertIdx:]
 		}
-		// Remove possible `explain` prefix.
-		bindSQL = bindSQL[insertIdx:]
-		return strings.Replace(bindSQL, "SELECT", fmt.Sprintf("SELECT /*+ %s*/", planHint), 1)
+	default:
+		logutil.Logger(ctx).Debug("[sql-bind] unexpected statement type when generating bind SQL", zap.Any("statement", stmtNode))
+		return ""
 	}
-	logutil.Logger(ctx).Debug("[sql-bind] unexpected statement type when generating bind SQL", zap.Any("statement", stmtNode))
-	return ""
+
+	bindSQL := getBindSQL(stmtNode, planHints, defaultDB, blockStartOffset)
+	if bindSQL == "" {
+		return ""
+	}
+	return afterProcessFunc(bindSQL)
+}
+
+func getBindSQL(stmtNode ast.StmtNode, planHints []*ast.TableOptimizerHint, defaultDB string, blockOffsetStart int) string {
+	counter := -1
+	blockHintProcessor := hint.BlockHintProcessor{}
+	offsetToHintMap := make(map[int][]*ast.TableOptimizerHint)
+	for _, pHint := range planHints {
+		hintOffset := blockHintProcessor.ParseOffsetFromBlockName(pHint.QBName, blockOffsetStart)
+		hints := offsetToHintMap[hintOffset]
+		if hints == nil {
+			hints = make([]*ast.TableOptimizerHint, 0)
+		}
```

Comment:
Useless. The `nil slice` can append directly.

Reply:
done